exercises/exercise_session_1.tex

```diff
@@ -0,0 +1,57 @@
+
+At this point, we established a method to ensure \emph{reliable} communication. As a result, we are able to coordinate the entities to ensure the correct behaviour of the distributed system. Sadly, reliable \emph{point-to-point} communication is probably not sufficient in a distributed system since more high level features are probably required. For example, let us assume processes $p$, and $q$, where $p$ needs to send message $m$ to $q$. Of course, $p$ could send the message directly to $q$ using our reliable point-to-point abstraction. However, this requires a \emph{direct} connection from $p$ to $q$. It is trivial to see that this the point-to-point protocol assumes some abstraction, i.e., a direct connection between $p$ and $q$ is available. As we will see in later courses, distributed systems, some of which you may or may not use (e.g., BitTorrent), usually place additional layers of abstraction on top of existing infrastructure (albeit with the same functionality) to achieve a specific goal, while at the same time, remove the abstraction when something efficient needs to be done.\\
+
+In this exercise session, we will study and apply some of these abstractions in order to captivate their importance, and to what end they can be utilized.
+
+\section{Exercises}
+\label{sec:exercises}
+
+\subsection*{Exercise 1}
+\label{sec:exercise_1}
+
+\emph{Define liveness and safety formally. Additionally, specify an example for both liveness and safety (the examples in the lecture and the book do not count).}
+
+\subsection*{Exercise 2}
+\label{sec:exercise_2}
+
+\emph{In class we saw that liveness and safety can be employed to prove the correctness of a distributed system. Let us assume a traditional producer-consumer setting. Which of the following are liveness properties and why?}
+
+\begin{enumerate}
+\item A consumer will eventually consume an element from the queue.
+\item Two consumers cannot consume the same head item concurrently.
+\item A producer has to stop producing if the queue is full.
+\item After a producer stops, it will start again if the queue is not full.
+\end{enumerate}
+
+\subsection*{Exercise 3}
+\label{sec:exercise_3}
+
+\emph{What is the problem vector clocks solved compared to logical clocks? How do they operate differently? In what type of timing abstraction are vector clocks employed?}
```

Comment:
Other interesting question: what is the main limitation of vector clocks? (it grows with the number of processes)